### ISSUE TITLE
Out of order notifications

### ISSUE THREAD

**Issue by cconstab** [atsign-foundation/at_client_sdk#608]
**Describe the bug**
When using atTalk ([attalk](https://github.com/cconstab/at_talk)) if there is no await on the notification  on line 181 (in branch 'testing-await') then notifications appear out of order. Using the await statement things are in order but VERY slow.

**To Reproduce**
1. First I clone the repo and move to the right branch
2. Then I send a file using unix pipes (see below)
3. with the await on line 181 in place everything arrives in order, without it order is not kept
4. I am pretty sure @gkc and I noticed this in testing a while back and Gary put in a fix ? Not sure why it is happening again/now.

**Expected behavior**
Never get out of order notifications, with await or not on the sending side of the notification.

**Screenshots**
With await in place 

```
cconstab@servalan:~/at_talk$ cat pubspec.yaml | dart bin/at_talk.dart -a "@iot_device02" -t "@iot_device01"
Synching your data....................
name: at_talk
description: A sample command-line application.
version: 1.0.0
# homepage: https://www.example.com
environment:
  sdk: '>=2.17.3 <3.0.0'
dependencies: 
  at_client: ^3.0.3
  at_onboarding_cli: ^1.0.0
  chalkdart: ^2.0.8
dev_dependencies:
  lints: ^1.0.0
  test: ^1.16.0
^C
cconstab@servalan:~/at_talk$

```

```
cconstab@servalan:~/at_talk$ dart bin/at_talk.dart -a "@iot_device01" -t "@iot_device02"
Synching your data..................
@iot_device02: name: at_talk
@iot_device02: description: A sample command-line application.
@iot_device02: version: 1.0.0
@iot_device02: # homepage: https://www.example.com
@iot_device02: environment:
@iot_device02: sdk: '>=2.17.3 <3.0.0'
@iot_device02: dependencies:
@iot_device02: at_client: ^3.0.3
@iot_device02: at_onboarding_cli: ^1.0.0
@iot_device02: chalkdart: ^2.0.8
@iot_device02: dev_dependencies:
@iot_device02: lints: ^1.0.0
@iot_device02: test: ^1.16.0
@iot_device01: ^C
cconstab@servalan:~/at_talk$
```

With await removed

```
cconstab@servalan:~/at_talk$ dart bin/at_talk.dart -a "@iot_device01" -t "@iot_device02"
Synching your data.........................
@iot_device02: at_client: ^3.0.3
@iot_device02: chalkdart: ^2.0.8
@iot_device02: description: A sample command-line application.
@iot_device02: name: at_talk
@iot_device02: lints: ^1.0.0
@iot_device02: # homepage: https://www.example.com
@iot_device02: at_onboarding_cli: ^1.0.0
@iot_device02: sdk: '>=2.17.3 <3.0.0'
@iot_device02: test: ^1.16.0
@iot_device02: environment:
@iot_device02: dependencies:
@iot_device02: version: 1.0.0
@iot_device02: dev_dependencies:
@iot_device01: 
```

Note lines received out of order (e.g. 'name: at_talk'

**Smartphone (please complete the following information):**
`Dart SDK version: 2.17.6 (stable) (Unknown timestamp) on "linux_x64"`

**Were you using an @‎application when the bug was found?**
  - atTalk

**Additional context**

**Comment by cconstab** [comment]
Demo video

https://user-images.githubusercontent.com/6131216/179156613-27622c88-8178-48ee-97bc-f4dbe2c273cc.mp4

**Comment by gkc** [comment]
@cconstab no I've not ever done any changes client-side re notification order

do the notifications leave the client in the wrong order, or does the ordering problem happen after that?

**Comment by gkc** [comment]
Have you tried doing an await just for the very first one you send, and no awaits for any subsequent ones?

**Comment by cconstab** [comment]
I have not but will do and report back... I remember testing this in the past and we saw the same thing and then changed something..

**Comment by cconstab** [comment]
I will get the de TLS code in place and see for sure what is going on on the wire

**Comment by gkc** [comment]
Maybe this is the Dart event loop working as intended? Tracing the code through, there are multiple places where we call `await` in the course of sending the notification. Each of those is an opportunity for the event loop to start handling another event (e.g. another notification being sent) and then there's the opportunity for all those many subsequent events to be processed out of order (as far as the human expectation goes)

**Comment by gkc** [comment]
But we need to see precisely where the order is being lost. Could be in the sender client; could be in the sender's secondary; could be in the receiver's secondary when it is sending to the monitor)

**Comment by cconstab** [comment]
@gkc from what I am seeing at this point the notifications leave the client in order but we need to set up a test rig where we can see all the notifications, in particular on the secondaries..

**Comment by gkc** [comment]
Yeah I'll do that tomorrow

**Comment by gkc** [comment]
@cconstab The notifications are definitely not in order when leaving the sender client in my test rig. What I'm seeing is a semi-random order when alice client is sending `notify` commands to alice secondary, but then the order is perfectly preserved in alice secondary, bob secondary, and bob client

I've done a few things to make it easier to see what's going on. I'm investigating in more detail in the client now to see what's happening. I'll push an at_client_sdk branch and an at_talk branch later

**Comment by gkc** [comment]
#### Observations:
* Every time an await goes to do I/O especially network I/O, the time to do the I/O is not consistent. This allows for events to overtake each other
* When we send the very first chat message, a few things happen - looking up secondary address, fetching shared key, creating it if it's not there, etc
  * We also call `isNetworkAvailable()` on every single request
* The various calls to `isNetworkAvailable()` for example cannot be guaranteed to complete in the order they were made, as they are dependent on things in the operating system, the networking stack, and the network itself

#### Experiments:
If I use `await` when doing the very first notification
then all subsequent notifications arrive in something very close to, but not quite, the correct order

If I use `await` when doing the very first notification **AND** I also remove the call to `isNetworkAvailable()` from `AtClientImpl.notifyChange(..)`
then all subsequent notifications arrive in the correct order every time

#### Recommendations / next steps
* We need to discuss the general use of `isNetworkAvailable()` before we make calls which will go over the network. I think this is utterly pointless because (a) the network might be available when we check, but unavailable a microsecond later when we try to write to a socket (b) we'll discover pretty fast *anyway* if the network is unavailable when we try to write to a socket. Network availability is a state that is of interest to user and may help them understand why a request failed, not something to be checked before every interaction, imo
* Note that when using await, much delay is introduced by `NotificationServiceImpl.notify(...)` because, once it's sent the notification (via `AtClientImpl.notifyChange()`), it then proceeds to poll the server for the notification status, every 2 seconds, until the status is neither empty nor 'queued'.  We need to discuss this area of the client functionality as this surprised me and it definitely feels wrong.
* Finally - while this pattern will preserve notification order rather well for this particular client, it is not a guarantee of order preservation for anything else. For that to be true we will need to introduce an outbound notification queue in the client so that order is re-enforced when finally sending the command to the secondary

**Comment by gkc** [comment]
@cconstab I've put my branch here https://github.com/gkc/at_talk/tree/OutOfOrderNotifications

**Comment by cconstab** [comment]
@gkc I looked at your branch and it works fine once I comment out the ` if (!await NetworkUtil.isNetworkAvailable()) {` stuff in at_client_impl.dart.

Do we need to fan out this ticket into three parts to track the issues you  found ?

**Comment by gkc** [comment]
Yeah I'll fan out some other issues. I'll close this one once I've published the at_client change which removes the call to isNetworkAvailable

**Comment by gkc** [comment]
PS pubspec in my forked at_talk has a dependencies override to my branch in at_client

**Comment by gkc** [comment]
I've addressed all three points above in #615

**Comment by gkc** [comment]
Will be in at_client release 3.0.31. In the meanwhile, at_talk is using dependency override to at_client trunk
